src/attribution/onFCP.ts

```diff
@@ -19,42 +19,40 @@ import {getLoadState} from '../lib/getLoadState.js';
 import {getNavigationEntry} from '../lib/getNavigationEntry.js';
 import {isInvalidTimestamp} from '../lib/isInvalidTimestamp.js';
 import {onFCP as unattributedOnFCP} from '../onFCP.js';
-import {
-  FCPMetric,
-  FCPMetricWithAttribution,
-  FCPReportCallback,
-  FCPReportCallbackWithAttribution,
-  ReportOpts,
-} from '../types.js';
+import {FCPMetric, FCPMetricWithAttribution, ReportOpts} from '../types.js';
+
+const attributeFCP = (metric: FCPMetric): FCPMetricWithAttribution => {
+  const metricWithAttribution = metric as FCPMetricWithAttribution;
 
-const attributeFCP = (metric: FCPMetric): void => {
   if (metric.entries.length) {
     const navigationEntry = getNavigationEntry();
     const fcpEntry = metric.entries[metric.entries.length - 1];
 
     if (navigationEntry) {
       const responseStart = navigationEntry.responseStart;
-      if (isInvalidTimestamp(responseStart)) return;
+      // TODO(bckenny): this is wrong.
+      if (isInvalidTimestamp(responseStart)) return metricWithAttribution;
```

Comment:
I didn't notice this until making this change, but since `attribution` is required, isn't this a bug to not have any `attribution` defined in this case?

Reply:
That is interesting how did the cast work below if this is required?:

```js
(metric as FCPMetricWithAttribution).attribution = {
```

Surely when the initial case, and before `attribution` is set it's a `FCPMetricWithAttribution` type but without an `attribution`?

Just curious btw rather than saying you're not right here...

Reply:
> Surely when the initial case, and before `attribution` is set it's a `FCPMetricWithAttribution` type but without an `attribution`?

Exactly. The cast/[type assertion](https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#type-assertions) overrides the type checker, so the compiler believes it's an `FCPMetricWithAttribution` and so will tell you `attribution` is defined, but it might not be. Type assertions are an escape hatch when you know something is correct (or it won't matter for this particular code) but the compiler can't tell for whatever reason. The downside is that the human half of that has to catch any mistakes because the compiler is coming into it with intentionally incorrect assumptions.

Typescript in general isn't great at typing objects that evolve in shape over time like this. Usually you'd do something like make `attribution` optional in type-land or you could create a new object (`{...metric, attribution}`) in value-land and it would track that, but that brings in JS we don't want to run in this case.

I think either we want optional `attribution` for these two cases (if there really is no attribution for them) or use the same kind of fallback empty attribution like is used a little farther down.

Reply:
> Ahhh, good catch

@philipwalton should these get the default `attribution` objects?

Reply:
I think they should be an empty object in these cases. I think it would be confusing if the attribution property itself was required on some metrics but optional on others.

Reply:
> I think they should be an empty object in these cases. I think it would be confusing if the attribution property itself was required on some metrics but optional on others (and since in v3 it was required on all metrics, I think we should stick with that).

Yeah, that's what I was assuming.